mrp_subcontracting_skip_no_negative/models/stock_move.py

```diff
@@ -31,7 +34,43 @@ def _action_done(self, cancel_backorder=False):
             ):
                 continue
             moves_with_no_check -= move
-        res = super(StockMove, self - moves_with_no_check)._action_done(
+            # If you have not been able to allocate previously it is because there is
+            # no stock, therefore it will leave the stock negative, we deduct the
+            # quantity checking the components and show the corresponding error.
+            test_condition = (
+                config["test_enable"] and self.env.context.get("test_stock_no_negative")
+            ) or not config["test_enable"]
+            if not test_condition:
+                continue
+            qty_precision = self.env["decimal.precision"].precision_get(
+                "Product Unit of Measure"
+            )
+            for p_move in unassigned_productions.move_raw_ids.filtered(
+                lambda x: x.state != "assigned"
+                and not x.product_id.allow_negative_stock
+                and not x.product_id.categ_id.allow_negative_stock
+                and not x.location_id.allow_negative_stock
+            ):
+                product = p_move.product_id.sudo()
+                location = p_move.location_id
+                location_qty = product.with_context(location=location.id).free_qty
+                new_qty = location_qty - p_move.product_uom_qty
+                if float_compare(new_qty, 0, precision_digits=qty_precision) == -1:
+                    raise ValidationError(
+                        _(
+                            "You cannot validate this stock operation because the "
+                            "stock level of the component product '{name}' would become "
+                            "negative ({qty}) on the stock location '{location}' and "
+                            "negative stock is not allowed for this product and/or "
+                            "location."
+                        ).format(
+                            name=product.display_name,
+                            qty=new_qty,
+                            location=location.complete_name,
+                        )
+                    )
+        moves_with_check = self - moves_with_no_check
+        res = super(StockMove, moves_with_check.with_context(no_backorder=True))._action_done(
```

Comment:
I guess the current code doesn't handle the case well where partial receipt is processed for a picking which only contains subcontracted products.

```suggestion
        # Assign `no_backorder=True` context to the first batch of moves for _action_done()
        # to avoid generation of a backorder during the process, in case there are two batches
        # of moves (with negative check and without) to be processed in _action_done().
        if moves_with_no_check != self:
            moves_with_no_check = moves_with_no_check.with_context(no_backorder=True)
        res = super(StockMove, moves_with_no_check)._action_done(
            cancel_backorder=cancel_backorder
        )
        res += super(StockMove, self - moves_with_no_check)._action_done(
            cancel_backorder=cancel_backorder
        )
```